### PR TITLE
Split IModifiedFileEntry into notebook and textfile

### PR DESCRIPTION
**Pull request by DonJayamanne**
# Summary of changes
* Split IModifiedFileEntry into IModifiedTextFileEntry and IModifiedNotebookFileEntry
    * Enforce the use of the interface (previously some parts of the code referenced the concrete `ChatEditingModifiedFileEntry` class directly
* Creating, serializing, deserializing of snapshots moved into a separate interfaces and classes
    * Notebook snapshots are not same as text snapshots are we deal with binary content
    * The operations are async, as a result, creating, deserializing, restoring snapshots are now `async`
    * The above `async` change can be seen in a number of places
    * The `TextSnapshotEntry` class keeps all of the creation, serialization, de-serialization in one place (previously in a few other files)
* Separate interface for Notebook entry files
    * Allows for separate implementation of accepting agent edits, snapshots & the like.

**Finally**
* I have removed changes to notebook files to keep this PR small (hence the compilation errors)
* The plan is to ensure most of the code is behind an `if` condition to ensure none of the new code paths are hit until we're ready to enable them. I.e. `IModifiedNotebookFileEntry` will not be used for a while in production.

### CODE REVIEW

File: src/vs/workbench/contrib/notebook/browser/contrib/chatEdit/notebookSynchronizer.ts [hunk 1]
```diff
@@ -249,7 +249,7 @@ export class NotebookModelSynchronizer extends Disposable {
 	private _originalModel?: Promise<NotebookTextModel>;
 	private async getOriginalModel(entry: IModifiedFileEntry): Promise<NotebookTextModel> {
 		if (!this._originalModel) {
-			this._originalModel = this.originalModelRefFactory.getOrCreate(entry, this.model.viewType).then(ref => this._register(ref).object);
+			this._originalModel = this.originalModelRefFactory.getOrCreate(entry as unknown as any, this.model.viewType).then(ref => this._register(ref).object);
```
Comment (DonJayamanne) [inline src/vs/workbench/contrib/notebook/browser/contrib/chatEdit/notebookSynchronizer.ts:252]: This file will be deleted in a separate PR, hence the dirty cast to avoid compilation errors.
Wanted to minize diffs in this PR hence not removed here.

File: src/vs/workbench/contrib/chat/common/chatEditingService.ts [hunk 2]
```diff
@@ -110,19 +118,134 @@ export const enum ChatEditingSessionChangeType {
...
+	readonly originalModel: INotebookTextModel;
+	readonly modifiedModel: INotebookTextModel;
+	readonly diffInfo: IObservable<ICellDiffInfo[]>;
+	acceptAgentCellEdits(cellUri: URI, textEdits: TextEdit[], isLastEdits: boolean): void;
+	acceptAgentNotebookEdits(edits: ICellEditOperation[], isLastEdits: boolean): Promise<void>;
+	createSnapshot(requestId: string | undefined): Promise<INotebookSnapshotEntry>;
+	restoreFromSnapshot(snapshot: INotebookSnapshotEntry): Promise<void>;
+}
+
+export type IModifiedFileEntry = IModifiedTextFileEntry | IModifiedNotebookFileEntry;
```
Comment (DonJayamanne) [inline src/vs/workbench/contrib/chat/common/chatEditingService.ts:247]: This is the crux of the PR, the split into two distinct interfaces, one for text and one for notebooks

Comment (DonJayamanne) [reply src/vs/workbench/contrib/chat/common/chatEditingService.ts:247]: Separate PR to create class implementing the notebook interface

File: src/vs/workbench/contrib/chat/browser/notebook.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * This function will be removed, but left here for now to avoid breaking changes.
+ */
+export function isNotebookChatSupported() {
+	return false;
+}
+
+/**
+ * This function will be removed, but left here for now to avoid breaking changes.
+ */
+export function isNotebook(_resource: URI): boolean {
```
Comment (DonJayamanne) [inline src/vs/workbench/contrib/chat/browser/notebook.ts:18]: This is to leave backwards compatibility, ensure none of the new code paths are hit now.

File: src/vs/workbench/contrib/chat/common/chatEditingService.ts [hunk 4]
```diff
@@ -38,7 +46,7 @@ export interface IChatEditingService {
 
 	startOrContinueEditingSession(chatSessionId: string): Promise<IChatEditingSession>;
 	getEditingSession(resource: URI): IChatEditingSession | null;
-	createSnapshot(requestId: string): void;
+	createSnapshot(requestId: string): Promise<void>;
```
Comment (DonJayamanne) [inline src/vs/workbench/contrib/chat/common/chatEditingService.ts:49]: We cannot sequentially create notebook snapshots.

**Comment by DonJayamanne** [comment]
Got a few merge conflicts, hence removing from review.

File: src/vs/workbench/contrib/chat/browser/chatContentParts/chatMarkdownContentPart.ts [hunk 5]
```diff
@@ -378,7 +378,7 @@ class CollapsedCodeBlock extends Disposable {
 			if (!isStreaming && isComplete) {
 				const labelAdded = this.element.querySelector('.label-added') ?? this.element.appendChild(dom.$('span.label-added'));
 				const labelRemoved = this.element.querySelector('.label-removed') ?? this.element.appendChild(dom.$('span.label-removed'));
-				const changes = modifiedEntry?.diffInfo.read(r);
+				const changes = isTextFileEntry(modifiedEntry) ? modifiedEntry?.diffInfo.read(r) : undefined;
 				if (changes && !changes?.identical && !changes?.quitEarly) {
```
Comment (DonJayamanne) [inline src/vs/workbench/contrib/chat/browser/chatContentParts/chatMarkdownContentPart.ts:382]: This functionality will be added later.

**Review by 818Nawaf — APPROVED** [review]
